xtask/src/test.rs

```diff
@@ -0,0 +1,216 @@
+use clap::Parser;
+use flate2::read::GzDecoder;
+use indicatif::{ProgressBar, ProgressStyle};
+use tar::Archive;
+use xshell::{cmd, Shell};
+
+use crate::tempdir::TempDir;
+
+const ARCHITEST_VERSION: &str = "0.4";
+
+#[derive(Debug, Parser)]
+pub(crate) struct Options {
+    /// Target architecture. It can be different than the host.
+    #[clap(long, default_value = "x86_64-unknown-linux-musl")]
+    target: String,
+
+    /// Preserve the temporary directory with architest artifacts (relevant
+    /// only for cross builds).
+    #[arg(long)]
+    preserve_tempdir: bool,
+
+    /// Build and run the release target
+    #[clap(long)]
+    release: bool,
+
+    /// Use architest/qemu even for a native target.
+    #[clap(long)]
+    force_architest: bool,
+}
+
+fn download_tarball<P>(url: &str, tarball_path: P) -> Result<()>
+where
+    P: AsRef<Path> + std::fmt::Debug,
+{
+    let mut response = reqwest::blocking::get(url)?;
+    let content_length = response
+        .content_length()
+        .ok_or(anyhow::anyhow!("Failed to get content length of {url}"))?;
+
+    let pb = ProgressBar::new(content_length).with_message("Downloading architest archive");
+    pb.set_style(ProgressStyle::default_bar()
+        .template("{spinner:.green} [{elapsed_precise}] [{wide_bar:.cyan/blue}] {bytes}/{total_bytes} ({eta})")?
+        .progress_chars("#>-"));
+
+    let mut file = File::create(tarball_path)?;
+    let mut buffer = vec![0; 4096];
+    let mut downloaded: u64 = 0;
+
+    while downloaded < content_length {
+        let bytes_read = response.read(&mut buffer)?;
+        if bytes_read == 0 {
+            break;
+        }
+        file.write_all(&buffer[..bytes_read])?;
+        downloaded += bytes_read as u64;
+        pb.set_position(downloaded);
+    }
+
+    pb.finish_with_message("Download complete");
+
+    Ok(())
+}
+
+fn extract_tarball<P>(tempdir: &TempDir, tarball_path: P) -> Result<()>
+where
+    P: AsRef<Path>,
+{
+    let file = File::open(tarball_path)?;
+    let decoder = GzDecoder::new(file);
+    let mut archive = Archive::new(decoder);
+    archive.unpack(tempdir)?;
+
+    Ok(())
+}
+
+fn download_and_unpack_architest(tempdir: &TempDir, architest_tarball: &str) -> Result<()> {
+    let url =
+        format!("https://github.com/exein-io/architest/releases/download/{ARCHITEST_VERSION}/{architest_tarball}");
+    let tarball_path = tempdir.join(architest_tarball);
+
+    download_tarball(&url, &tarball_path)?;
+    extract_tarball(tempdir, &tarball_path)?;
+
+    Ok(())
+}
+
+fn test_architest(sh: Shell, options: Options, binary_file: &str) -> Result<()> {
+    let Options {
+        target,
+        preserve_tempdir,
+        ..
+    } = options;
+
+    let tempdir = TempDir::new("pulsar-architest", preserve_tempdir)?;
+    sh.change_dir(&tempdir);
+
+    let (architest_tarball, qemu_cmd, qemu_args) = match target.as_str() {
+        "aarch64-unknown-linux-musl" => (
+            "aarch64_6.6.tar.gz",
```

Comment:
we have to test few kernel version, not only the `6.6`

Reply:
Done, all versions should be covered now.